### ISSUE TITLE
Umbraco deploy timeout issue

### ISSUE THREAD

**Issue by Deba22** [umbraco/Umbraco.Deploy.Issues#179]
Hi All,

We have migrated our project to umbraco version 10.5.1 from version 8. Im trying to do a partial restore of one node from Live to Staging. The partial restore fails with following error:

```
d3800087 - Executing, 10%
d3800087 - Executing, 10% - Get remote session
d3800087 - Executing, 80% - Exception!
d3800087 - Executing, 80% - Abort remote session
d3800087 - Executing, 80% - Restore caches and indexes
d3800087 - Failed, 80% - Total time 00:07:41.263.

-- EXCEPTION ---------------------------------------------------
 (proper exception traces at the end of this log)
Umbraco.Deploy.Infrastructure.Exceptions.RemoteApiTimeoutException: The remote API returned content indicating a platform timeout.
----------------------------------------------------------------

EXCEPTION:

Umbraco.Deploy.Infrastructure.Exceptions.RemoteApiTimeoutException: The remote API returned content indicating a platform timeout.
   at Umbraco.Deploy.Infrastructure.Http.DeployHttpClient.SucceedOrThrow(HttpResponseMessage result, CancellationToken token)
   at Umbraco.Deploy.Infrastructure.Http.DeployHttpClient.SendAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken token)
   at Umbraco.Deploy.Infrastructure.Http.DeployHttpClient.SendAsync(HttpRequestMessage request, CancellationToken token)
   at Umbraco.Deploy.Infrastructure.Environments.RemoteUmbracoEnvironment.BeginSessionAsync(Guid sessionId, CancellationToken token)
   at Umbraco.Deploy.Infrastructure.Work.WorkItems.DeployRestoreWorkItemBase.GetSessionOnWithBatchInfo(IUmbracoEnvironment env, Nullable`1 batchIndex, Nullable`1 numberOfBatches)
   at Umbraco.Deploy.Infrastructure.Work.WorkItems.DeployRestoreWorkItemBase.GetSessionOn(IUmbracoEnvironment env)
   at Umbraco.Deploy.Infrastructure.Work.WorkItems.TargetPartialRestoreWorkItem.ExecuteAsyncSub(IWorkContext context, List`1 resume, CancellationToken token)
   at Umbraco.Deploy.Infrastructure.Work.WorkItems.TargetPartialRestoreWorkItem.ExecuteAsyncSub(IWorkContext context, List`1 resume, CancellationToken token)
   at Umbraco.Deploy.Infrastructure.Work.WorkItems.TargetPartialRestoreWorkItem.ExecuteAsync(IWorkContext context, CancellationToken token)

REMOTE:
TEXT: &lt;html&gt;&lt;head&gt;&lt;title&gt;500 - The request timed out.&lt;/title&gt;&lt;/head&gt;&lt;body&gt;  &lt;font color ="#aa0000"&gt;         &lt;h2&gt;500 - The request timed out.&lt;/h2&gt;&lt;/font&gt;  The web server failed to respond within the specified time.&lt;/body&gt;&lt;/html&gt;
```

I have the following app settings in appsettings.json file
```
    "Deploy": {
      "Settings": {
        "SessionTimeout": "0.0:40:00",
        "SourceDeployTimeout": "0.0:40:00",
        "DatabaseCommandTimeout": "0.0:40:00",
        "HttpClientTimeout": "0.0:40:00",
        "DiskOperationsTimeout": "0.0:25:00"
      },
```
Even though I have set and increased the timeout values, the partial restore fails well within 10 mins

Here are the versions of Umbraco deploy I'm using:
Umbraco.Deploy.Contrib => 10.1.1
Umbraco.Deploy.Forms => 10.2.3
Umbraco.Deploy.OnPrem => 10.2.3

Please let me know if more information is required.

**Comment by AndyButland** [comment]
We've detailed some options you can look at on the "troubleshooting" page of the documentation, which you can find [here](https://docs.umbraco.com/umbraco-deploy/troubleshooting#slow-responses-or-timeouts-when-restoring-or-transferring).  In particular it here may well be worth looking at the step where we discuss "Ensure signatures are pre-cached".

**Comment by ronaldbarendse** [comment]
We've finally been able to find the exact cause of this issue and have fixed the remote timeout issue in Deploy 13.3.2, 14.2.2 and 15.0.2.

For those that are curious: Deploy uses server-to-server communication between environments to transfer or restore content, media or other items (initiated via the backoffice). This already uses separate `Begin*`, `WaitIdle` and `End*` endpoints to handle long-running tasks, where e.g. the `BeginCreateManifest` endpoint queues/starts a `Task` (which can be used to asynchronously run the action), `WaitIdle` returns when it is completed and `EndCreateManifest` returns the result of the completed task. However, there was one major flaw in the implementation: the actions were running synchronously, resulting in running the whole action in the `Begin*` endpoint and queueing an already completed task. If the action then takes more than the maximum request time (Azure Web Apps has a hard limit of 4 minutes), the connection is closed and the deployment fails. We're now forcing all actions to run asynchronously using [`await Task.Yield();`](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.yield?view=net-8.0).
